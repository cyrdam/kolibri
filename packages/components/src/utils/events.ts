// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
