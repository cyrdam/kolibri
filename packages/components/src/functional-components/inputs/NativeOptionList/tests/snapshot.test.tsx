import { h } from '@stencil/core';
import { renderFunctionalComponentToSpecPage } from '../../../../utils/testing';
import NativeOptionListFc from '../NativeOptionList';

describe('NativeOptionListFc', () => {
	it('should render and match snapshot', async () => {
		const options = [{ label: 'Option 1', value: '1' }];
		const page = await renderFunctionalComponentToSpecPage(() => <NativeOptionListFc options={options} />);
		expect(page.root).toMatchSnapshot();
	});

	it('should render with no options', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <NativeOptionListFc options={[]} />);
		expect(page.root).toBeNull();
	});

	it('should render with a single option', async () => {
		const options = [{ label: 'Option 1', value: '1' }];
		const page = await renderFunctionalComponentToSpecPage(() => <NativeOptionListFc options={options} />);
		expect(page.root).toMatchSnapshot();
	});

	it('should render with multiple options', async () => {
		const options = [
			{ label: 'Option 1', value: '1' },
			{ label: 'Option 2', value: '2' },
		];
		const page = await renderFunctionalComponentToSpecPage(() => <NativeOptionListFc options={options} />);
		expect(page.root).toMatchSnapshot();
	});

	it('should render with nested option groups', async () => {
		const options = [
			{
				label: 'Group 1',
				options: [
					{ label: 'Option 1', value: '1' },
					{ label: 'Option 2', value: '2' },
				],
			},
		];
		const page = await renderFunctionalComponentToSpecPage(() => <NativeOptionListFc options={options} />);
		expect(page.root).toMatchSnapshot();
	});
});
