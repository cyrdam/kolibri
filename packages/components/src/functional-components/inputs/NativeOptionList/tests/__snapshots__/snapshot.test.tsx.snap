// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeOptionListFc should render and match snapshot 1`] = `
<option value="-0">
  Option 1
</option>
`;

exports[`NativeOptionListFc should render with a single option 1`] = `
<option value="-0">
  Option 1
</option>
`;

exports[`NativeOptionListFc should render with multiple options 1`] = `
<option value="-0">
  Option 1
</option>
`;

exports[`NativeOptionListFc should render with nested option groups 1`] = `
<optgroup label="Group 1">
  <option value="-0-0">
    Option 1
  </option>
  <option value="-0-1">
    Option 2
  </option>
</optgroup>
`;
